### PR TITLE
Coefficient generation fixes for IMEX adaptivity

### LINKED ISSUE #516: Fix GPU adaptivity on IMEX solve
**Describe the bug**
GPU is not advancing in time when adaptivity and IMEX are enabled.  This could be an issue with the coefficient matrix change implemented that does not compute the 1D coefficient matrix at every timestep.

### PR DESCRIPTION
**Pull request by ckendrick**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

- This introduces adaptivity for IMEX, and adds temporary changes to fix coefficient generation to use the adapted level rather than each level up to `max_level`.
- Fixes issue with 1D transformation in IMEX when running with adaptivity enabled.
- Moves several redundant `generate_all_coefficients` calls in IMEX when not running problem with a collision operator.
- Hash table and degrees of freedom are now saved to file
- Fixes several realspace transformation issues for plotting, file output due to adaptivity grid changes.
- Fixes redundant realspace transformation in `main` even if not plotting or saving realspace data.

Closes #516.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [x] Bugfix
- [x] New feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## What systems has this change been tested on?

Ubuntu 20.04

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/moment.cpp [hunk 1]
```diff
@@ -38,19 +39,20 @@ void moment<P>::createFlist(PDE<P> const &pde, options const &opts)
...
 
-  if (this->vector.empty() || opts.do_adapt_levels())
+  if (this->vector.empty() || opts.do_adapt_levels)
   {
     distribution_plan const plan = get_plan(get_num_ranks(), hash_table);
     auto rank                    = get_rank();
     int const degree             = pde.get_dimensions()[0].get_degree();
     auto tmp = combine_dimensions(degree, hash_table, plan.at(rank).row_start,
                                   plan.at(rank).row_stop, this->fList[0]);
+    this->vector = fk::vector<P>();
```
Comment (mkstoyanov) [inline src/moment.cpp]: Why create a new empty vector only to immediately resize it. We can just construct it with the right size.

Comment (mkstoyanov) [reply src/moment.cpp]: Look further down, we are moving into vector. This is dumb, just move tmp into vector.

There was a bug that prevented moving into a vector with different size, but we fixed that awhile back.

File: src/time_advance.cpp [hunk 2]
```diff
@@ -581,8 +616,21 @@ imex_advance(PDE<P> &pde, matrix_list<P> &operator_matrices,
   // Create rho_2s
   fk::vector<P> mom0(dense_size);
   fm::gemv(pde.moments[0].get_moment_matrix(), x, mom0);
+
+  /*
+  fk::matrix<P> tmp = asgard::operator_two_scale<P>(degree, level);
```
Comment (mkstoyanov) [inline src/time_advance.cpp]: Pointless code, we can remove it or we should have comment describing what it does.

File: src/time_advance.cpp [hunk 3]
```diff
@@ -581,8 +616,21 @@ imex_advance(PDE<P> &pde, matrix_list<P> &operator_matrices,
...
+  fk::vector<P> mom0_real_coeffs(dense_size);
+  fm::gemv(tmp, mom0, mom0_real_coeffs);
+
+  fm::scal(static_cast<P>((1.0 / (std::sqrt(4.0 * PI * std::pow(2, level))))),
+  mom0_real_coeffs); mom0_real_coeffs.print("mom0_real_coeffs");
+  */
+
   fk::vector<P> &mom0_real = pde.moments[0].create_realspace_moment(
       pde_1d, mom0, adaptive_grid_1d.get_table(), transformer, tmp_workspace);
+  // mom0_real.print("mom0_real");
```
Comment (mkstoyanov) [inline src/time_advance.cpp]: More commented code.

File: src/moment.cpp (hunk 1 above)
Comment (ckendrick) [reply src/moment.cpp]: This empty vector was accidentally left over from something else.

I'll update the following two lines now that the vector move bug is fixed.

File: src/time_advance.cpp (hunk 2 above)
Comment (ckendrick) [reply src/time_advance.cpp]: Sorry, this is not needed and should have been removed.

File: src/asgard_dimension.hpp [hunk 4]
```diff
@@ -331,14 +350,27 @@ struct dimension
...
+  {
+    expect(level >= 0);
+
+    if (level >= static_cast<int>(this->mass_.size()))
+    {
+      this->mass_.resize(level + 1);
+    }
+    this->mass_[level].clear_and_resize(new_mass.nrows(), new_mass.ncols());
+    expect(this->mass_[level].nrows() == new_mass.nrows());
+    this->mass_[level] = std::move(new_mass);
```
Comment (quantumsteve) [inline src/asgard_dimension.hpp]:
```suggestion
    this->mass_[level] = std::move(new_mass);
```

**Review by mkstoyanov — DISMISSED** [review]

File: src/time_advance.cpp [hunk 5]
```diff
@@ -473,8 +498,16 @@ imex_advance(PDE<P> &pde, matrix_list<P> &operator_matrices,
...
+    std::cout << " dim0 lev = " << level << "\n";
+    std::cout << " dim1 lev = " << pde.get_dimensions()[1].get_level() << "\n";
     for (auto &m : pde.moments)
     {
+      // m.createFlist(pde, program_opts);
+      // expect(m.get_fList().size() > 0);
+
+      // m.createMomentVector(pde, program_opts, adaptive_grid.get_table());
+      // expect(m.get_vector().size() > 0);
+
```
Comment (quantumsteve) [inline src/time_advance.cpp]:
```suggestion
```

File: src/time_advance.cpp [hunk 6]
```diff
@@ -440,7 +458,14 @@ imex_advance(PDE<P> &pde, matrix_list<P> &operator_matrices,
...
   PDE pde_1d = PDE(pde, PDE<P>::extract_dim0);
-  adapt::distributed_grid adaptive_grid_1d(pde_1d, program_opts);
+
+  options const opts_1d = make_options(
+      {"-d 3", "-f",
+       "-l " + std::to_string(pde.get_dimensions()[0].get_level())});
+  adapt::distributed_grid adaptive_grid_1d(pde_1d, opts_1d);
+
+  // asgard::basis::wavelet_transform<P, asgard::resource::host> const
+  // transformer_1d(opts_1d, pde_1d, false);
```
Comment (quantumsteve) [inline src/time_advance.cpp]:
```suggestion

```

File: src/time_advance_tests.cpp [hunk 7]
```diff
@@ -1494,6 +1501,160 @@ TEMPLATE_TEST_CASE("IMEX time advance - twostream", "[imex]", double)
...
+  TestType E_kin_initial = 0.0;
+
+  // number of DOF for the FG case: (degree * 2^level)^2 = 9.216e3
+  int const fg_dof = std::pow(degree * std::pow(2, levels[0]), 2);
+
+  // -- time loop
+  for (auto i = 0; i < opts.num_time_steps; ++i)
+  {
+    std::cout.setstate(std::ios_base::failbit);
+    auto const time          = i * pde->get_dt();
```
Comment (quantumsteve) [inline src/time_advance_tests.cpp]: give this an explicit type (double?)

File: src/asgard_dimension.hpp (hunk 4 above)
Comment (ckendrick) [reply src/asgard_dimension.hpp]: Using just `std::move` here causes this assertion to fail: https://github.com/project-asgard/asgard/blob/24ec80cc4061804d90f69fa039c86e8208510198/src/tensors.hpp#L1757

Comment (mkstoyanov) [reply src/asgard_dimension.hpp]: Line 1757 refers to copy assignment. Are you sure it isn't line 1933, which is the move assignment (and it is wrong).

Comment (ckendrick) [reply src/asgard_dimension.hpp]: Using only the `std::move` at line 363 (commenting out 361 and 362), here is the back trace in gdb:
```
#4  0x00005555555f199b in asgard::fk::matrix<double, (asgard::mem_type)0, (asgard::resource)0>::operator= (this=0x555584aaf430, a=...)
    at ~/asgard/src/tensors.hpp:1757
#5  0x00005555555f1add in asgard::dimension<double>::set_mass_matrix (this=0x555560407af8, new_mass=..., level=4)
    at ~/asgard/src/pde/../asgard_dimension.hpp:363
#6  0x00005555555ecb5e in asgard::PDE<double>::update_dimension_mass_mat (this=0x555584aa7310, dim_index=1, mass=..., level=4)
    at ~/asgard/src/pde/pde_base.hpp:815
```

Comment (mkstoyanov) [reply src/asgard_dimension.hpp]: That's odd, line 1757 is the copy-assignment, somehow the compiler isn't moving but it is copying.

I'll have to look into it tonight.

Comment (mkstoyanov) [reply src/asgard_dimension.hpp]: Oh, you're passing in a const matrix and you cannot move a const matrix.

If you want to do this, you need to change the call to `set_mass_matrix()` to accept r-value-ref and move when you pass in the matrix.

Comment (ckendrick) [reply src/asgard_dimension.hpp]: Ah yes, thanks for catching that! I rebased again now that your move assignment fix #595 is in and updated `set_mass_matrix()` to properly use move assignment now.

**Review by mkstoyanov — APPROVED** [review]

**Comment by mkstoyanov** [comment]
This looks OK, but in a future PR we should reduce some of the I/O noise. Maybe have a CMake or CLI option to make it output less details, the `cout` statements will be particularly problematic when we move to MPI and multiple ranks.

**Comment by ckendrick** [comment]
Yes I agree, we can disable those in a future PR.

**Pull request merged**